File: Semester-3/Homework-1/LazyCalculation/LazyFactory.cs
using System;

namespace LazyCalculation
{
    /// <summary>
    /// Lazy calculation object creator
    /// </summary>
    public static class LazyFactory<T>
    {
        /// <summary>
        /// Creates a single thread lazy calculation function
        /// </summary>
        /// <param name="supplier">Supplier function</param>
        public static LazySingleThread<T> CreateSingleThread(Func<T> supplier)
            => new LazySingleThread<T>(supplier);

        /// <summary>
        /// Creates a multi thread lazy calculation function
        /// </summary>
        /// <param name="supplier">Supplier function</param>
        public static LazyParallel<T> CreateMultiThread(Func<T> supplier)
            => new LazyParallel<T>(supplier);
    }
}

File: Semester-4/Homework-7/Lazy/LazyFactory.fs
namespace Lazy

/// Type for Lazy factory
type LazyFactory () =
    
    /// Creates a single thread version of Lazy
    static member CreateSingleThreadLazy<'a>(supplier : unit -> 'a) =
        new LazySingleThread<'a>(supplier) :> ILazy<'a>

    /// Creates a locking multi thread version of Lazy
    static member CreateParallelLazy<'a>(supplier : unit -> 'a) =
        new LazyParallel<'a>(supplier) :> ILazy<'a>

    /// Creates a lock-free multi thread version of Lazy
    static member CreateLockFreeLazy<'a>(supplier : unit -> 'a) =
        new LazyLockFree<'a>(supplier) :> ILazy<'a>